aquadoggo/src/network/transport.rs

```diff
@@ -0,0 +1,57 @@
+// SPDX-License-Identifier: AGPL-3.0-or-later
+
+use futures::future::Either;
+use libp2p::core::muxing::StreamMuxerBox;
+use libp2p::core::transport::upgrade::Version;
+use libp2p::core::transport::{Boxed, OrTransport};
+use libp2p::identity::Keypair;
+use libp2p::noise::NoiseAuthenticated;
+use libp2p::yamux::YamuxConfig;
+use libp2p::{quic, relay, PeerId, Transport};
+
+// Build the transport stack to be used by the network swarm
+pub async fn build_transport(
+    key_pair: &Keypair,
+    relay_client_enabled: bool,
+) -> (
+    Boxed<(PeerId, StreamMuxerBox)>,
+    Option<relay::client::Behaviour>,
+) {
+    // Create QUIC transport (provides transport, security and multiplexing in a single protocol).
+    // The QUIC transport will also be used for any relayed connections
+    let quic_config = quic::Config::new(key_pair);
+    let quic_transport = quic::tokio::Transport::new(quic_config);
+
+    let (transport, relay_client) = if relay_client_enabled {
+        // Generate a relay transport and client behaviour if relay client mode is enabled
+        let (relay_transport, relay_client) = relay::client::new(key_pair.public().to_peer_id());
+        let relay_client = Some(relay_client);
+
+        // Add encryption and multiplexing to the relay transport
+        let relay_transport = relay_transport
+            .upgrade(Version::V1)
+            .authenticate(NoiseAuthenticated::xx(key_pair).unwrap())
+            .multiplex(YamuxConfig::default());
```

Comment:
Looking at https://docs.rs/libp2p/latest/libp2p/relay/client/struct.Transport.html I understand that this whole block can be removed now, right?

Reply:
Incorrect. The relayed connection still needs to be secured and multiplexed.

Take a look half-way through my comment here (where I write: "And the second suggestion (do not make the relay transport optional) results in a "mismatched types" error:"):

https://github.com/libp2p/rust-libp2p/discussions/3653#discussioncomment-5391803

Reply:
Ah, thank you! I'm still not sure if I understand what is happening here though. How is the relay transport encrypted and multiplexed if it is not used in the end?

I read the comment in https://github.com/libp2p/rust-libp2p/discussions/3766:

> Relay itself isn't a transport. A relayed connection will always run on the connection that you establish to the relay

And that makes me wonder why at all this needs to be set then if all is handled by QUIC in the end?

Reply:
Or in other words: Are we just setting these things to satisfy the API but actually relay transport is _not_ encrypted and multiplexed in these lines of code?

Reply:
@sandreae has asked about this : https://github.com/libp2p/rust-libp2p/discussions/3766#discussioncomment-5595346

Reply:
This is an ongoing convo with the rust-libp2p devs, so let's not hold this PR up anymore :+1: